hal/halx86/acpi/madt.c

```diff
@@ -12,77 +13,113 @@
 #include <debug.h>
 
+#define APIC_MODE_LEGACY 0x10
+
 /* GLOBALS ********************************************************************/
 
-PROCESSOR_IDENTITY HalpStaticProcessorIdentity[MAXIMUM_PROCESSORS] = {{0}};
-PPROCESSOR_IDENTITY HalpProcessorIdentity = NULL;
 HALP_APIC_INFO_TABLE HalpApicInfoTable;
-ACPI_TABLE_MADT *MadtTable;
-ACPI_SUBTABLE_HEADER *AcpiHeader;
-ACPI_MADT_LOCAL_APIC *LocalApic;
+
+static PROCESSOR_IDENTITY HalpStaticProcessorIdentity[MAXIMUM_PROCESSORS];
+PPROCESSOR_IDENTITY HalpProcessorIdentity = HalpStaticProcessorIdentity;
 
 /* FUNCTIONS ******************************************************************/
 
 VOID
 HalpParseApicTables(
     _In_ PLOADER_PARAMETER_BLOCK LoaderBlock)
 {
+    ACPI_TABLE_MADT *MadtTable;
+    ACPI_SUBTABLE_HEADER *AcpiHeader;
     ULONG_PTR TableEnd;
-    ULONG ValidProcessorCount;
-
-    /* We only support legacy APIC for now, this will be updated in the future */
-    HalpApicInfoTable.ApicMode = 0x10;
-    MadtTable = HalAcpiGetTable(LoaderBlock, 'CIPA');
 
-    AcpiHeader = (ACPI_SUBTABLE_HEADER*)MadtTable;
-    AcpiHeader->Length = sizeof(ACPI_TABLE_MADT);
-    TableEnd = (ULONG_PTR)MadtTable + MadtTable->Header.Length;
+    MadtTable = HalAcpiGetTable(LoaderBlock, APIC_SIGNATURE);
+    if (!MadtTable)
+    {
+        DPRINT1("HalAcpiGetTable(APIC) failed\n");
+        return;
+    }
 
-    HalpApicInfoTable.ProcessorCount = 0;
-    HalpProcessorIdentity = HalpStaticProcessorIdentity;
+    /* TODO: We only support legacy APIC for now */
+    HalpApicInfoTable.ApicMode = APIC_MODE_LEGACY;
 
-    AcpiHeader = (ACPI_SUBTABLE_HEADER*)MadtTable;
-    AcpiHeader->Length = sizeof(ACPI_TABLE_MADT);
+    AcpiHeader = (ACPI_SUBTABLE_HEADER *)((ULONG_PTR)MadtTable + sizeof(*MadtTable));
     TableEnd = (ULONG_PTR)MadtTable + MadtTable->Header.Length;
+    DPRINT("MadtTable %p, subtables %p - %08Ix\n", MadtTable, AcpiHeader, TableEnd);
 
-    while ((ULONG_PTR)AcpiHeader <= TableEnd)
+    while ((ULONG_PTR)AcpiHeader < TableEnd)
     {
-        LocalApic = (ACPI_MADT_LOCAL_APIC*)AcpiHeader;
-
-        if (LocalApic->Header.Type == ACPI_MADT_TYPE_LOCAL_APIC &&
-            LocalApic->Header.Length == sizeof(ACPI_MADT_LOCAL_APIC))
+        switch (AcpiHeader->Type)
         {
-            ValidProcessorCount = HalpApicInfoTable.ProcessorCount;
-
-            HalpProcessorIdentity[ValidProcessorCount].LapicId = LocalApic->Id;
-            HalpProcessorIdentity[ValidProcessorCount].ProcessorId = LocalApic->ProcessorId;
-
-            HalpApicInfoTable.ProcessorCount++;
-
-            AcpiHeader = (ACPI_SUBTABLE_HEADER*)((ULONG_PTR)AcpiHeader + AcpiHeader->Length);
-        }
-        else
-        {
-            /* End the parsing early if we don't use the currently selected table */
-            AcpiHeader = (ACPI_SUBTABLE_HEADER*)((ULONG_PTR)AcpiHeader + 1);
+            case ACPI_MADT_TYPE_LOCAL_APIC:
+            {
+                ACPI_MADT_LOCAL_APIC *LA = (ACPI_MADT_LOCAL_APIC *)AcpiHeader;
```

Comment:
I wouldn't mind if you added bounds checking, seems like it would be good practice. For example the `AcpiHeader < TableEnd` check doesn't actually guarantee that a `ACPI_SUBTABLE_HEADER` (or any of the larger structures) will fit.
E.g. here you'd want `NT_ASSERT((ULONG_PTR)(LA + 1) <= TableEnd)` or equivalent.
Also think about what would happen if `AcpiHeader->Length` was 0.

If any of those fail, asserting would be good if we know the debugger works, otherwise I don't have a strong preference between bugcheck and simply returning with incomplete data.

Reply:
Checks added.
As a first step, I used `return` only, to at least allow to review the rest of the PR.
Do you think separating into 2-pass would be better: first validate table structure (fail=abort), then deal with actual data (fail=skip)?